util/process_wrapper/flags.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::error::Error;
+use std::fmt;
+use std::fmt::Write;
+use std::mem::{replace, take};
+
+#[derive(Debug, Clone)]
+pub(crate) enum FlagParseError {
+    UnkownFlag(String),
+    RequiredArgumentMissing(String),
+    ValueMissing(String),
+    ProvidedMultipleTimes(String),
+}
+
+impl fmt::Display for FlagParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::UnkownFlag(ref flag) => write!(f, "unknown flag \"{}\"", flag),
+            Self::RequiredArgumentMissing(ref flag) => {
+                write!(f, "flag \"{}\" is required but was not provided", flag)
+            }
+            Self::ValueMissing(ref flag) => write!(f, "flag \"{}\" missing parameter(s)", flag),
+            Self::ProvidedMultipleTimes(ref flag) => {
+                write!(f, "flag \"{}\" can only appear once", flag)
+            }
+        }
+    }
+}
+impl Error for FlagParseError {}
+
+struct FlagDef<'a, T> {
+    name: String,
+    required: bool,
+    help: String,
+    data: &'a mut T,
+}
+
+impl<'a, T> fmt::Display for FlagDef<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let required = if self.required { "(required)" } else { "" };
+        write!(f, "{} {}\t{}", self.name, required, self.help)
+    }
+}
+
+impl<'a, T> fmt::Debug for FlagDef<'a, T> {
```

Comment:
Why not `#[derive(Debug)]`?

Reply:
Because I don't want to print the reference'd data.